File: android/core/src/test/kotlin/io/github/notkanaa/equipe/core/logic/LogicTestSupport.kt
package io.github.notkanaa.equipe.core.logic

import io.github.notkanaa.equipe.core.AppCalendar
import io.github.notkanaa.equipe.core.AppError
import io.github.notkanaa.equipe.core.AssignmentEvent
import io.github.notkanaa.equipe.core.LocalNotification
import io.github.notkanaa.equipe.core.NotificationAuthorization
import io.github.notkanaa.equipe.core.NotificationScheduler
import io.github.notkanaa.equipe.core.NowProvider
import io.github.notkanaa.equipe.core.RealtimeEvent
import io.github.notkanaa.equipe.core.RealtimeService
import io.github.notkanaa.equipe.core.TaskDraft
import io.github.notkanaa.equipe.core.TaskItem
import io.github.notkanaa.equipe.core.TaskPriority
import io.github.notkanaa.equipe.core.TaskService
import io.github.notkanaa.equipe.core.TaskStatus
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger
import kotlin.time.Duration

// Test doubles and fixtures of the logic suites (port of Tests/TeamTasksCoreTests/Logic/LogicTestSupport.swift).
// Swift's manual test clock is replaced by kotlinx-coroutines-test's virtual time (runTest, advanceTimeBy,
// runCurrent); `LogicSleepers` counts the pending waits like `LogicTestClock.sleeperCount`.

// region Fixtures

object LogicFixtures {
    val paris: ZoneId = ZoneId.of("Europe/Paris")
    val parisCalendar: AppCalendar = AppCalendar.frenchGregorian(paris)

    val me: UUID = UUID.fromString("00000000-0000-0000-0000-00000000000A")
    val other: UUID = UUID.fromString("00000000-0000-0000-0000-00000000000B")
    val groupA: UUID = UUID.fromString("00000000-0000-0000-0000-0000000000A1")
    val groupB: UUID = UUID.fromString("00000000-0000-0000-0000-0000000000B2")

    /** A date in Europe/Paris. */
    fun date(year: Int, month: Int, day: Int, hour: Int = 0, minute: Int = 0, second: Int = 0): Instant =
        ZonedDateTime.of(year, month, day, hour, minute, second, 0, paris).toInstant()

    /** Deterministic UUID from a small number. */
    fun uuid(number: Int): UUID =
        UUID.fromString("10000000-0000-0000-0000-" + Integer.toHexString(number).padStart(12, '0'))

    fun task(
        number: Int,
        title: String? = null,
        group: UUID = groupA,
        status: TaskStatus = TaskStatus.TODO,
        priority: TaskPriority = TaskPriority.MEDIUM,
        due: Instant? = null,
        createdAt: Instant = date(2026, 1, 1),
        completedAt: Instant? = null,
        assignees: List<UUID> = listOf(me),
        groupName: String? = "Coloc' rue des Lilas",
    ): TaskItem = TaskItem(
        id = uuid(number),
        groupId = group,
        title = title ?: "Tâche $number",
        status = status,
        priority = priority,
        dueAt = due,
        createdBy = other,
        createdAt = createdAt,
        updatedAt = createdAt,
        completedAt = if (status == TaskStatus.DONE) completedAt ?: createdAt else null,
        assigneeIds = assignees,
        groupName = groupName,
    )
}

// endregion

// region Waiting

/** Counts the coroutines currently waiting in a `sleep` (Swift `LogicTestClock.sleeperCount`); time stays virtual. */
class LogicSleepers {
    private val active = AtomicInteger()

    val count: Int get() = active.get()

    val sleep: suspend (Duration) -> Unit = { duration ->
        active.incrementAndGet()
        try {
            delay(duration)
        } finally {
            active.decrementAndGet()
        }
    }
}

// endregion

// region Notification scheduler

/** Thrown by [LogicFakeScheduler.add] for its failing ids. */
class LogicSchedulingFailure : Exception("échec simulé")

class LogicFakeScheduler(
    authorization: NotificationAuthorization = NotificationAuthorization.AUTHORIZED,
) : NotificationScheduler {
    private val lock = Any()
    private var authorizationValue = authorization
    private val pendingById = HashMap<String, LocalNotification>()
    private val addedLog = ArrayList<LocalNotification>()
    private val removeLog = ArrayList<List<String>>()
    private var failing: Set<String> = emptySet()
    private var checks = 0

    var authorization: NotificationAuthorization
        get() = synchronized(lock) { authorizationValue }
        set(value) = synchronized(lock) { authorizationValue = value }

    /** Ids whose `add` throws. */
    var failingIds: Set<String>
        get() = synchronized(lock) { failing }
        set(value) = synchronized(lock) { failing = value }

    /** Pending (scheduled, not delivered) notifications by id. */
    val pending: Map<String, LocalNotification> get() = synchronized(lock) { HashMap(pendingById) }
    val pendingIds: List<String> get() = synchronized(lock) { pendingById.keys.sorted() }

    /** Every successful `add`, in order. */
    val added: List<LocalNotification> get() = synchronized(lock) { addedLog.toList() }
    val removeCalls: List<List<String>> get() = synchronized(lock) { removeLog.toList() }
    val authorizationChecks: Int get() = synchronized(lock) { checks }

    /** Simulates an already-pending request (e.g. scheduled by a previous launch). */
    fun seedPending(notification: LocalNotification) {
        synchronized(lock) { pendingById[notification.id] = notification }
    }

    fun resetCallLog() {
        synchronized(lock) {
            addedLog.clear()
            removeLog.clear()
        }
    }

    override suspend fun authorizationStatus(): NotificationAuthorization = synchronized(lock) {
        checks += 1
        authorizationValue
    }

    override suspend fun requestAuthorization(): Boolean =
        synchronized(lock) { authorizationValue == NotificationAuthorization.AUTHORIZED }

    override suspend fun pendingIdentifiers(prefix: String): List<String> =
        synchronized(lock) { pendingById.keys.filter { it.startsWith(prefix) }.sorted() }

    override suspend fun add(notification: LocalNotification) {
        synchronized(lock) {
            if (notification.id in failing) throw LogicSchedulingFailure()
            addedLog.add(notification)
            // "Deliver now" requests are shown immediately and never stay pending.
            if (notification.fireDate != null) {
                pendingById[notification.id] = notification
            }
        }
    }

    override suspend fun removePending(ids: List<String>) {
        synchronized(lock) {
            removeLog.add(ids.toList())
            for (id in ids) pendingById.remove(id)
        }
    }

    override suspend fun setBadge(count: Int) {}
}

/**
 * Scheduler whose n-th `add` call (1-based) can be held until the test opens it, to reproduce interleavings of
 * concurrent callers. A held request is registered when released (like a late platform call).
 */
class LogicGatedScheduler(private val gatedCalls: Set<Int>) : NotificationScheduler {
    private val lock = Any()
    private val pendingById = HashMap<String, LocalNotification>()
    private var addCalls = 0
    private val waiting = HashMap<Int, CompletableDeferred<Unit>>()

    val pending: Map<String, LocalNotification> get() = synchronized(lock) { HashMap(pendingById) }

    /** Calls currently held. */
    val waitingCalls: Set<Int> get() = synchronized(lock) { waiting.keys.toSet() }
    val addCallCount: Int get() = synchronized(lock) { addCalls }

    /** Releases the held call [call]. */
    fun open(call: Int) {
        val gate = synchronized(lock) { waiting.remove(call) }
        gate?.complete(Unit)
    }

    override suspend fun authorizationStatus(): NotificationAuthorization = NotificationAuthorization.AUTHORIZED

    override suspend fun requestAuthorization(): Boolean = true

    override suspend fun pendingIdentifiers(prefix: String): List<String> =
        synchronized(lock) { pendingById.keys.filter { it.startsWith(prefix) }.sorted() }

    override suspend fun add(notification: LocalNotification) {
        val gate = synchronized(lock) {
            addCalls += 1
            if (addCalls in gatedCalls) CompletableDeferred<Unit>().also { waiting[addCalls] = it } else null
        }
        gate?.await()
        synchronized(lock) { pendingById[notification.id] = notification }
    }

    override suspend fun removePending(ids: List<String>) {
        synchronized(lock) {
            for (id in ids) pendingById.remove(id)
        }
    }

    override suspend fun setBadge(count: Int) {}
}

// endregion

// region Task service

class LogicFakeTaskService : TaskService {
    private val lock = Any()
    private val tasksById = HashMap<UUID, TaskItem>()
    private var events: List<AssignmentEvent> = emptyList()
    private var error: AppError? = null
    private val lookupLog = ArrayList<UUID>()
    private val sinceLog = ArrayList<Instant>()

    fun put(task: TaskItem) {
        synchronized(lock) { tasksById[task.id] = task }
    }

    fun removeTask(id: UUID) {
        synchronized(lock) { tasksById.remove(id) }
    }

    /**
     * Server-side assignment rows returned by `assignments(since)` (filtered on `assignedAt > since` only, so tests can
     * check that the notifier also ignores self-assignments).
     */
    var assignmentEvents: List<AssignmentEvent>
        get() = synchronized(lock) { events }
        set(value) = synchronized(lock) { events = value.toList() }

    var assignmentsError: AppError?
        get() = synchronized(lock) { error }
        set(value) = synchronized(lock) { error = value }

    val lookups: List<UUID> get() = synchronized(lock) { lookupLog.toList() }
    val sinceCalls: List<Instant> get() = synchronized(lock) { sinceLog.toList() }

    override suspend fun task(id: UUID): TaskItem = synchronized(lock) {
        lookupLog.add(id)
        tasksById[id] ?: throw AppError.NotFound
    }

    override suspend fun assignments(since: Instant): List<AssignmentEvent> = synchronized(lock) {
        sinceLog.add(since)
        error?.let { throw it }
        events.filter { it.assignedAt > since }.sortedBy { it.assignedAt }
    }

    override suspend fun tasks(groupId: UUID, includeOldDone: Boolean): List<TaskItem> = unused()

    override suspend fun myTasks(includeDone: Boolean): List<TaskItem> = unused()

    override suspend fun create(groupId: UUID, draft: TaskDraft): TaskItem = unused()

    override suspend fun update(taskId: UUID, draft: TaskDraft): TaskItem = unused()

    override suspend fun setStatus(taskId: UUID, status: TaskStatus): TaskItem = unused()

    override suspend fun delete(taskId: UUID): Unit = unused()

    private fun unused(): Nothing = throw AppError.Unknown("non utilisé")
}

// endregion

// region Realtime service

/**
 * Cold fake of the realtime service: collecting [events] registers a subscription, leaving the collection
 * (completion, failure or cancellation) terminates it.
 */
class LogicFakeRealtime : RealtimeService {
    private class Subscription(val userId: UUID, val groupIds: List<UUID>) {
        val channel = Channel<RealtimeEvent>(Channel.UNLIMITED)

        @Volatile
        var terminated = false
    }

    private val lock = Any()
    private val subscriptions = ArrayList<Subscription>()

    override fun events(userId: UUID, groupIds: List<UUID>): Flow<RealtimeEvent> = flow {
        val subscription = Subscription(userId, groupIds.toList())
        synchronized(lock) { subscriptions.add(subscription) }
        try {
            for (event in subscription.channel) {
                emit(event)
            }
        } finally {
            subscription.terminated = true
            subscription.channel.close()
        }
    }

    val subscriptionCount: Int get() = synchronized(lock) { subscriptions.size }

    fun groupIds(index: Int): List<UUID> = synchronized(lock) { subscriptions[index].groupIds }

    fun userId(index: Int): UUID = synchronized(lock) { subscriptions[index].userId }

    fun isTerminated(index: Int): Boolean = synchronized(lock) { subscriptions[index].terminated }

    /** Sends an event on the latest subscription (dropped when it is terminated). */
    fun send(event: RealtimeEvent) {
        latest()?.channel?.trySend(event)
    }

    /** Ends the latest stream from the server side (connection lost). */
    fun finishLatest() {
        latest()?.channel?.close()
    }

    /** Fails the latest stream (e.g. a network error surfaced by the client library). */
    fun failLatest(error: Throwable) {
        latest()?.channel?.close(error)
    }

    private fun latest(): Subscription? = synchronized(lock) { subscriptions.lastOrNull() }
}

// endregion

// region Group ids provider

class LogicGroupIdsSource(ids: List<UUID>) {
    private val lock = Any()
    private var idsValue = ids
    private var callCount = 0
    private var failing = false

    var ids: List<UUID>
        get() = synchronized(lock) { idsValue }
        set(value) = synchronized(lock) { idsValue = value }

    var fails: Boolean
        get() = synchronized(lock) { failing }
        set(value) = synchronized(lock) { failing = value }

    val calls: Int get() = synchronized(lock) { callCount }

    fun fetch(): List<UUID> = synchronized(lock) {
        callCount += 1
        if (failing) throw AppError.Network
        idsValue
    }
}

/** Collects forwarded assignments. */
class LogicAssignmentSink {
    private val lock = Any()
    private val log = ArrayList<RealtimeAssignment>()

    val received: List<RealtimeAssignment> get() = synchronized(lock) { log.toList() }

    fun receive(assignment: RealtimeAssignment) {
        synchronized(lock) { log.add(assignment) }
    }
}

/** Mutable "now" shared with the code under test. */
class LogicNow(value: Instant) {
    @Volatile
    var value: Instant = value

    val provider: NowProvider = { this.value }
}

// endregion

File: Packages/TeamTasksKit/Tests/TeamTasksCoreTests/Logic/LogicTestSupport.swift
import Foundation
import Testing
@testable import TeamTasksCore

// Test doubles and fixtures for the Logic suites. Names are prefixed with `Logic` so they never clash with
// helpers of other suites of this test target.

// MARK: - Fixtures

enum LogicFixtures {
    static let paris = TimeZone(identifier: "Europe/Paris")!
    static let parisCalendar = Calendar.frenchGregorian(timeZone: paris)

    static let me = UUID(uuidString: "00000000-0000-0000-0000-00000000000A")!
    static let other = UUID(uuidString: "00000000-0000-0000-0000-00000000000B")!
    static let groupA = UUID(uuidString: "00000000-0000-0000-0000-0000000000A1")!
    static let groupB = UUID(uuidString: "00000000-0000-0000-0000-0000000000B2")!

    /// A date in Europe/Paris.
    static func date(_ year: Int, _ month: Int, _ day: Int, _ hour: Int = 0, _ minute: Int = 0, _ second: Int = 0) -> Date {
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = day
        components.hour = hour
        components.minute = minute
        components.second = second
        return parisCalendar.date(from: components)!
    }

    /// Deterministic UUID from a small number.
    static func uuid(_ number: Int) -> UUID {
        let hex = String(number, radix: 16)
        let padded = String(repeating: "0", count: 12 - hex.count) + hex
        return UUID(uuidString: "10000000-0000-0000-0000-\(padded)")!
    }

    static func task(
        _ number: Int,
        title: String? = nil,
        group: UUID = groupA,
        status: TaskStatus = .todo,
        priority: TaskPriority = .medium,
        due: Date? = nil,
        createdAt: Date = LogicFixtures.date(2026, 1, 1),
        completedAt: Date? = nil,
        assignees: [UUID] = [me],
        groupName: String? = "Coloc' rue des Lilas"
    ) -> TaskItem {
        TaskItem(
            id: uuid(number),
            groupId: group,
            title: title ?? "Tâche \(number)",
            status: status,
            priority: priority,
            dueAt: due,
            createdBy: other,
            createdAt: createdAt,
            updatedAt: createdAt,
            completedAt: status == .done ? (completedAt ?? createdAt) : nil,
            assigneeIds: assignees,
            groupName: groupName
        )
    }
}

// MARK: - Waiting

enum LogicWait {
    /// Polls `condition` (yielding, then sleeping 1 ms between checks) until it holds; records an issue after
    /// `timeout`. Only used to wait for other tasks to run; never to let simulated time pass.
    @MainActor
    static func until(
        _ description: String = "condition",
        timeout: Duration = .seconds(5),
        sourceLocation: SourceLocation = #_sourceLocation,
        _ condition: () -> Bool
    ) async {
        let deadline = ContinuousClock.now.advanced(by: timeout)
        var attempts = 0
        while !condition() {
            if ContinuousClock.now >= deadline {
                Issue.record("Timed out waiting for \(description)", sourceLocation: sourceLocation)
                return
            }
            attempts += 1
            if attempts % 20 == 0 {
                try? await Task.sleep(nanoseconds: 1_000_000)
            } else {
                await Task.yield()
            }
        }
    }

    /// Lets already-enqueued jobs run (used before asserting that something did *not* happen).
    @MainActor
    static func settle() async {
        for _ in 0..<50 {
            await Task.yield()
        }
        try? await Task.sleep(nanoseconds: 2_000_000)
        for _ in 0..<50 {
            await Task.yield()
        }
    }
}

// MARK: - Test clock

/// Manually driven `Clock`: sleepers only resume when the test calls `advance(by:)`.
final class LogicTestClock: Clock, @unchecked Sendable {
    struct Instant: InstantProtocol, Hashable {
        var offset: Swift.Duration

        func advanced(by duration: Swift.Duration) -> Instant { Instant(offset: offset + duration) }
        func duration(to other: Instant) -> Swift.Duration { other.offset - offset }
        static func < (lhs: Instant, rhs: Instant) -> Bool { lhs.offset < rhs.offset }
    }

    private struct Sleeper {
        var deadline: Instant
        var continuation: CheckedContinuation<Void, any Error>
    }

    private let lock = NSLock()
    private var current = Instant(offset: .zero)
    private var sleepers: [Int: Sleeper] = [:]
    private var cancelledBeforeRegistration: Set<Int> = []
    private var nextId = 0

    var now: Instant { lock.withLock { current } }
    var minimumResolution: Swift.Duration { .zero }

    /// Number of tasks currently sleeping on this clock.
    var sleeperCount: Int { lock.withLock { sleepers.count } }

    func sleep(until deadline: Instant, tolerance: Swift.Duration?) async throws {
        let id = lock.withLock {
            nextId += 1
            return nextId
        }
        try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, any Error>) in
                lock.lock()
                if cancelledBeforeRegistration.remove(id) != nil {
                    lock.unlock()
                    continuation.resume(throwing: CancellationError())
                } else if deadline <= current {
                    lock.unlock()
                    continuation.resume()
                } else {
                    sleepers[id] = Sleeper(deadline: deadline, continuation: continuation)
                    lock.unlock()
                }
            }
        } onCancel: {
            lock.lock()
            if let sleeper = sleepers.removeValue(forKey: id) {
                lock.unlock()
                sleeper.continuation.resume(throwing: CancellationError())
            } else {
                cancelledBeforeRegistration.insert(id)
                lock.unlock()
            }
        }
    }

    /// Moves time forward and wakes the sleepers whose deadline is reached.
    func advance(by duration: Swift.Duration) {
        lock.lock()
        current = current.advanced(by: duration)
        let now = current
        let due = sleepers.filter { $0.value.deadline <= now }.sorted { $0.value.deadline < $1.value.deadline }
        for (id, _) in due {
            sleepers.removeValue(forKey: id)
        }
        lock.unlock()
        for (_, sleeper) in due {
            sleeper.continuation.resume()
        }
    }
}

// MARK: - Notification scheduler

final class LogicFakeScheduler: NotificationScheduler, @unchecked Sendable {
    struct Failure: Error {}

    private let lock = NSLock()
    private var _authorization: NotificationAuthorization
    private var _pending: [String: LocalNotification] = [:]
    private var _added: [LocalNotification] = []
    private var _removeCalls: [[String]] = []
    private var _failingIds: Set<String> = []
    private var _authorizationChecks = 0

    init(authorization: NotificationAuthorization = .authorized) {
        _authorization = authorization
    }

    var authorization: NotificationAuthorization {
        get { lock.withLock { _authorization } }
        set { lock.withLock { _authorization = newValue } }
    }

    /// Ids whose `add` throws.
    var failingIds: Set<String> {
        get { lock.withLock { _failingIds } }
        set { lock.withLock { _failingIds = newValue } }
    }

    /// Pending (scheduled, not delivered) notifications by id.
    var pending: [String: LocalNotification] { lock.withLock { _pending } }
    var pendingIds: [String] { lock.withLock { _pending.keys.sorted() } }
    /// Every successful `add`, in order.
    var added: [LocalNotification] { lock.withLock { _added } }
    var removeCalls: [[String]] { lock.withLock { _removeCalls } }
    var authorizationChecks: Int { lock.withLock { _authorizationChecks } }

    /// Simulates an already-pending request (e.g. scheduled by a previous launch).
    func seedPending(_ notification: LocalNotification) {
        lock.withLock { _pending[notification.id] = notification }
    }

    func resetCallLog() {
        lock.withLock {
            _added = []
            _removeCalls = []
        }
    }

    func authorizationStatus() async -> NotificationAuthorization {
        lock.withLock {
            _authorizationChecks += 1
            return _authorization
        }
    }

    func requestAuthorization() async -> Bool {
        lock.withLock { _authorization == .authorized }
    }

    func pendingIdentifiers(prefix: String) async -> [String] {
        lock.withLock { _pending.keys.filter { $0.hasPrefix(prefix) }.sorted() }
    }

    func add(_ notification: LocalNotification) async throws {
        try lock.withLock {
            if _failingIds.contains(notification.id) { throw Failure() }
            _added.append(notification)
            // "Deliver now" requests are shown immediately and never stay pending.
            if notification.fireDate != nil {
                _pending[notification.id] = notification
            }
        }
    }

    func removePending(ids: [String]) async {
        lock.withLock {
            _removeCalls.append(ids)
            for id in ids {
                _pending.removeValue(forKey: id)
            }
        }
    }

    func setBadge(_ count: Int) async {}
}

/// Scheduler whose n-th `add` call (1-based) can be held until the test opens it, to reproduce interleavings
/// of concurrent callers. A held request is registered when released (like a late UNUserNotificationCenter call).
final class LogicGatedScheduler: NotificationScheduler, @unchecked Sendable {
    private let lock = NSLock()
    private var _pending: [String: LocalNotification] = [:]
    private var addCalls = 0
    private let gatedCalls: Set<Int>
    private var waiting: [Int: CheckedContinuation<Void, Never>] = [:]

    init(gatedCalls: Set<Int>) {
        self.gatedCalls = gatedCalls
    }

    var pending: [String: LocalNotification] { lock.withLock { _pending } }
    /// Calls currently held.
    var waitingCalls: Set<Int> { lock.withLock { Set(waiting.keys) } }
    var addCallCount: Int { lock.withLock { addCalls } }

    /// Releases the held call `call`.
    func open(_ call: Int) {
        let continuation = lock.withLock { waiting.removeValue(forKey: call) }
        continuation?.resume()
    }

    func authorizationStatus() async -> NotificationAuthorization { .authorized }
    func requestAuthorization() async -> Bool { true }

    func pendingIdentifiers(prefix: String) async -> [String] {
        lock.withLock { _pending.keys.filter { $0.hasPrefix(prefix) }.sorted() }
    }

    func add(_ notification: LocalNotification) async throws {
        let call = lock.withLock {
            addCalls += 1
            return addCalls
        }
        if gatedCalls.contains(call) {
            await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
                lock.withLock { waiting[call] = continuation }
            }
        }
        lock.withLock { _pending[notification.id] = notification }
    }

    func removePending(ids: [String]) async {
        lock.withLock {
            for id in ids {
                _pending.removeValue(forKey: id)
            }
        }
    }

    func setBadge(_ count: Int) async {}
}

// MARK: - Task service

final class LogicFakeTaskService: TaskService, @unchecked Sendable {
    private let lock = NSLock()
    private var _tasks: [UUID: TaskItem] = [:]
    private var _assignments: [AssignmentEvent] = []
    private var _assignmentsError: AppError?
    private var _lookups: [UUID] = []
    private var _sinceCalls: [Date] = []

    func put(_ task: TaskItem) {
        lock.withLock { _tasks[task.id] = task }
    }

    func removeTask(id: UUID) {
        lock.withLock { _tasks[id] = nil }
    }

    /// Server-side assignment rows returned by `assignments(since:)` (filtered on `assignedAt > since` only,
    /// so tests can check that the notifier also ignores self-assignments).
    var assignmentEvents: [AssignmentEvent] {
        get { lock.withLock { _assignments } }
        set { lock.withLock { _assignments = newValue } }
    }

    var assignmentsError: AppError? {
        get { lock.withLock { _assignmentsError } }
        set { lock.withLock { _assignmentsError = newValue } }
    }

    var lookups: [UUID] { lock.withLock { _lookups } }
    var sinceCalls: [Date] { lock.withLock { _sinceCalls } }

    func task(id: UUID) async throws -> TaskItem {
        try lock.withLock {
            _lookups.append(id)
            guard let task = _tasks[id] else { throw AppError.notFound }
            return task
        }
    }

    func assignments(since: Date) async throws -> [AssignmentEvent] {
        try lock.withLock {
            _sinceCalls.append(since)
            if let error = _assignmentsError { throw error }
            return _assignments.filter { $0.assignedAt > since }.sorted { $0.assignedAt < $1.assignedAt }
        }
    }

    func tasks(groupId: UUID, includeOldDone: Bool) async throws -> [TaskItem] { throw AppError.unknown("non utilisé") }
    func myTasks(includeDone: Bool) async throws -> [TaskItem] { throw AppError.unknown("non utilisé") }
    func create(groupId: UUID, draft: TaskDraft) async throws -> TaskItem { throw AppError.unknown("non utilisé") }
    func update(taskId: UUID, draft: TaskDraft) async throws -> TaskItem { throw AppError.unknown("non utilisé") }
    func setStatus(taskId: UUID, status: TaskStatus) async throws -> TaskItem { throw AppError.unknown("non utilisé") }
    func delete(taskId: UUID) async throws { throw AppError.unknown("non utilisé") }
    func addChecklistItem(taskId: UUID, title: String) async throws -> ChecklistItem { throw AppError.unknown("non utilisé") }
    func renameChecklistItem(itemId: UUID, title: String) async throws -> ChecklistItem { throw AppError.unknown("non utilisé") }
    func setChecklistItemDone(itemId: UUID, done: Bool) async throws -> ChecklistItem { throw AppError.unknown("non utilisé") }
    func deleteChecklistItem(itemId: UUID) async throws { throw AppError.unknown("non utilisé") }
    func completions(groupId: UUID, since: Date) async throws -> [TaskCompletion] { throw AppError.unknown("non utilisé") }
}

// MARK: - Realtime service

final class LogicFakeRealtime: RealtimeService, @unchecked Sendable {
    struct Subscription {
        var userId: UUID
        var groupIds: [UUID]
        var continuation: AsyncStream<RealtimeEvent>.Continuation
        var terminated = false
    }

    private let lock = NSLock()
    private var subscriptions: [Subscription] = []

    func events(userId: UUID, groupIds: [UUID]) -> AsyncStream<RealtimeEvent> {
        let (stream, continuation) = AsyncStream.makeStream(of: RealtimeEvent.self)
        let index = lock.withLock {
            subscriptions.append(Subscription(userId: userId, groupIds: groupIds, continuation: continuation))
            return subscriptions.count - 1
        }
        continuation.onTermination = { [weak self] _ in
            self?.lock.withLock { self?.subscriptions[index].terminated = true }
        }
        return stream
    }

    var subscriptionCount: Int { lock.withLock { subscriptions.count } }

    func groupIds(of index: Int) -> [UUID] { lock.withLock { subscriptions[index].groupIds } }
    func userId(of index: Int) -> UUID { lock.withLock { subscriptions[index].userId } }
    func isTerminated(_ index: Int) -> Bool { lock.withLock { subscriptions[index].terminated } }

    /// Sends an event on the latest subscription.
    func send(_ event: RealtimeEvent) {
        let continuation = lock.withLock { subscriptions.last?.continuation }
        continuation?.yield(event)
    }

    /// Ends the latest stream from the server side (connection lost).
    func finishLatest() {
        let continuation = lock.withLock { subscriptions.last?.continuation }
        continuation?.finish()
    }
}

// MARK: - Group ids provider

final class LogicGroupIdsSource: @unchecked Sendable {
    private let lock = NSLock()
    private var _ids: [UUID]
    private var _calls = 0
    private var _fails = false

    init(_ ids: [UUID]) {
        _ids = ids
    }

    var ids: [UUID] {
        get { lock.withLock { _ids } }
        set { lock.withLock { _ids = newValue } }
    }

    var fails: Bool {
        get { lock.withLock { _fails } }
        set { lock.withLock { _fails = newValue } }
    }

    var calls: Int { lock.withLock { _calls } }

    func fetch() throws -> [UUID] {
        try lock.withLock {
            _calls += 1
            if _fails { throw AppError.network }
            return _ids
        }
    }
}

/// Collects forwarded assignments.
final class LogicAssignmentSink: @unchecked Sendable {
    private let lock = NSLock()
    private var _received: [RealtimeAssignment] = []

    var received: [RealtimeAssignment] { lock.withLock { _received } }

    func receive(_ assignment: RealtimeAssignment) {
        lock.withLock { _received.append(assignment) }
    }
}

/// Mutable "now" shared with the code under test.
final class LogicNow: @unchecked Sendable {
    private let lock = NSLock()
    private var _value: Date

    init(_ value: Date) {
        _value = value
    }

    var value: Date {
        get { lock.withLock { _value } }
        set { lock.withLock { _value = newValue } }
    }

    var provider: NowProvider { { [self] in value } }
}
